### PR TITLE
Fixed #32260 handler500 as a Class-based view raises SystemCheckError

### PR DESCRIPTION
**Pull request by abbasidaniyal**
_check_custom_error_handlers in django/urls/resolver.py now checks if the handler is a function based view or class based view.
For class based view, num_parameters needs to be 2 as class based views take two parameters (self, request).

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -429,7 +430,10 @@ def _check_custom_error_handlers(self):
                 messages.append(Error(msg, hint=str(e), id='urls.E008'))
                 continue
             signature = inspect.signature(handler)
-            args = [None] * num_parameters
+            if signature.parameters.get('self'):
+                args = [None] * 2  # Class based views always take two arguements, (self, request).
+            else:
+                args = [None] * num_parameters
```
Comment (charettes) [inline django/urls/resolvers.py:436]: I think you want to allow an extra parameter when it's a class based view

In other words

```
# Class based views always take two arguements, (self, request).
```

Is not not true for some of these handlers.


```suggestion
            if signature.parameters.get('self'):
                num_parameters += 1
            args = [None] * num_parameters
```

**Review by charettes — COMMENTED** [review]
You should add tests for CBV handlers as well

File: django/urls/resolvers.py (hunk 1 above)
Comment (abbasidaniyal) [reply django/urls/resolvers.py:436]: Thank you for your suggestion!
Simply added +1 to the num_parameters will make the value of num_parameters for handler404, handler403 and handler400 as 3 which is not correct. 
https://github.com/django/django/blob/8d7085e0fd004af5431389f3d903aba6220d7957/django/urls/resolvers.py#L421

I can change it as `if signature.parameters.get('self') and status_code==500: ` but that would be the greatest solution in my opinion.

Comment (abbasidaniyal) [reply django/urls/resolvers.py:436]: > You should add tests for CBV handlers as well

Yes I will do that right away!

File: django/urls/resolvers.py [hunk 2]
```diff
@@ -429,7 +430,12 @@ def _check_custom_error_handlers(self):
                 messages.append(Error(msg, hint=str(e), id='urls.E008'))
                 continue
             signature = inspect.signature(handler)
-            args = [None] * num_parameters
+            if signature.parameters.get('self'):
+                args = [None] * 2  # Class based views always take two arguements, (self, request).
+                arguement_string = 'self, request'
```
Comment (jacobtylerwalls) [inline django/urls/resolvers.py:435]: "argument" is the spelling. :-)

Comment (abbasidaniyal) [reply django/urls/resolvers.py:435]: Oh my bad. 
Fixed it!

**Comment by adamchainz** [comment]
I think this ticket is going in the wrong direction and the check has really exposed a bug in the return value of `View.as_view()`.

`View.as_view()` returns a function that has signature `def view(request, *args, **kwargs):` - it does *not* take a `self` argument. The `self` is created per-request inside that `view()` function. However the `functools.update_wrapper()` calls inside `as_view()` copy the `self` argument to the `view()` function.

I think we should fix `View.as_view()` to drop the first positional argument (`self`) from the copied signature. Then the check will not need changing, and any other tools can see the true function signature.

**Comment by abbasidaniyal** [comment]
> I think this ticket is going in the wrong direction and the check has really exposed a bug in the return value of `View.as_view()`.
> 
> `View.as_view()` returns a function that has signature `def view(request, *args, **kwargs):` - it does _not_ take a `self` argument. The `self` is created per-request inside that `view()` function. However the `functools.update_wrapper()` calls inside `as_view()` incorrectly copy the function signature from the class methods, including their `self` argument, onto the `view()` function.
> 
> I think we should fix `View.as_view()` to drop the first positional argument (`self`) from the copied signature. Then the check will not need changing, and any other tools can see the true function signature.

That does make a lot of sense. Self should not be present on the signature in the first place!
I'll create a separate PR with the fix soon. Will close this as soon as a new one is created!

**Comment by felixxm** [comment]
@abbasidaniyal Thanks, feel-free to send a new PR.

**Comment by abbasidaniyal** [comment]
@felixxm Yes. I'm extremely sorry for delaying this. I'll be sending a new PR soon!

File: django/urls/resolvers.py (hunk 1 above)
Comment (lindacorbett) [reply django/urls/resolvers.py]: Yes I love SwissBorg tokens

**Pull request closed without merging**